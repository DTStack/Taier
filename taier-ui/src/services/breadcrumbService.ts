import { CATALOGUE_TYPE } from '@/constant';
import { TreeViewUtil } from '@dtinsight/molecule/esm/common/treeUtil';
import type molecule from '@dtinsight/molecule';
import type { UniqueId } from '@dtinsight/molecule/esm/common/types';
import type { IBreadcrumbItemProps } from '@dtinsight/molecule/esm/components';
import { catalogueService } from '.';

interface IBreadcrumbService {
    /**
     * 根据目录树 id 获取父节点并生成面包屑
     */
    getBreadcrumb: (id: UniqueId) => IBreadcrumbItemProps[];
}

export default class BreadcrumbService implements IBreadcrumbService {
    private hashTree: TreeViewUtil<molecule.model.IFolderTreeNodeProps> | null = null;
    constructor() {
        catalogueService.onUpdate(() => {
            this.updateTree();
        });
    }

    private updateTree = () => {
        const root = catalogueService.getRootFolder(CATALOGUE_TYPE.TASK);
        if (root) {
            this.hashTree = new TreeViewUtil<molecule.model.IFolderTreeNodeProps>(root);
        }
    };

    public getBreadcrumb = (id: UniqueId) => {
        if (this.hashTree) {
            const stack = [this.hashTree.getHashMap(id)];
            const res: IBreadcrumbItemProps[] = [];
            while (stack.length) {
                const hash = stack.pop();
                if (hash) {
                    res.push({
                        id: hash.node.id,
                        name: hash.node.name,
                    });

                    if (hash.parent) {
                        stack.push(this.hashTree.getHashMap(hash.parent));
                    }
                }
            }

            return res.reverse();
        }
        return [];
    };
}
