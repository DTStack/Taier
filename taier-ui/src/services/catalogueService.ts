import api from '@/api';
import { CATALOGUE_TYPE, MENU_TYPE_ENUM } from '@/constant';
import type { CatalogueDataProps } from '@/interface';
import { getTenantId, getUserId } from '@/utils';
import { fileIcon } from '@/utils/extensions';
import molecule from '@dtinsight/molecule';
import { GlobalEvent } from '@dtinsight/molecule/esm/common/event';
import { FileTypes, TreeNodeModel } from '@dtinsight/molecule/esm/model';
import { singleton } from 'tsyringe';
import functionManagerService from './functionManagerService';
import resourceManagerService from './resourceManagerService';

interface ICatalogueService {
    /**
     * 加载目录树的根目录
     */
    loadRootFolder: () => void;
    /**
     * 获取目录树的根目录
     */
    getRootFolder: (source: CATALOGUE_TYPE) => molecule.model.IFolderTreeNodeProps | undefined;
    /**
     * 获取目录树的子节点
     * @param node 更新当前 Node 节点的子节点，不改变 Node 节点
     */
    loadTreeNode: (node: Partial<Pick<CatalogueDataProps, 'id' | 'catalogueType'>>, source: CATALOGUE_TYPE) => void;
    /**
     * 目录树更新监听事件
     */
    onUpdate: (callback: () => void) => void;
}

export enum CatalogueEventKind {
    onUpdate = 'catalogue.update',
}

@singleton()
export default class CatalogueService extends GlobalEvent implements ICatalogueService {
    /**
     * 异步加载目录树节点
     */
    private getCatalogueViaNode = async (
        node: Partial<Pick<CatalogueDataProps, 'id' | 'catalogueType'>>
    ): Promise<CatalogueDataProps | undefined> => {
        if (!node) throw new Error('[getCatalogueViaNode]: failed to get catelogue');
        const res = await api.getOfflineCatalogue({
            isGetFile: true,
            nodePid: node.id,
            catalogueType: node.catalogueType,
            userId: getUserId(),
            tenantId: getTenantId(),
        });
        if (res.code === 1) {
            return res.data;
        }
        return undefined;
    };

    private getRootFolderViaSource = (data: CatalogueDataProps[], source: CATALOGUE_TYPE) => {
        switch (source) {
            case CATALOGUE_TYPE.TASK: {
                return data.find((item) => item.catalogueType === MENU_TYPE_ENUM.TASK);
            }

            case CATALOGUE_TYPE.RESOURCE: {
                return data.find((item) => item.catalogueType === MENU_TYPE_ENUM.RESOURCE);
            }

            case CATALOGUE_TYPE.FUNCTION: {
                return data.find((item) => item.catalogueType === MENU_TYPE_ENUM.FUNCTION);
            }
            default:
                return undefined;
        }
    };

    /**
     * Only transform current catalogue to treeNodeModel, ignore children
     */
    private transformCatalogueToTree = (catalogue: CatalogueDataProps | undefined, source: CATALOGUE_TYPE) => {
        if (!catalogue) return;
        const folderType = ['folder', 'catalogue'];
        switch (source) {
            case CATALOGUE_TYPE.RESOURCE: {
                const fileType = folderType.includes(catalogue.type) ? FileTypes.Folder : FileTypes.File;

                const id = fileType === FileTypes.File ? catalogue.id : `${catalogue.id}-folder`;
                return new TreeNodeModel({
                    // prevent same id between folder and file
                    id,
                    name: catalogue.name,
                    fileType,
                    icon: fileIcon(catalogue.resourceType, source),
                    isLeaf: fileType === FileTypes.File,
                    data: catalogue,
                    children: resourceManagerService.get(id)?.children || [],
                });
            }

            case CATALOGUE_TYPE.FUNCTION: {
                const { type, name } = catalogue;
                const fileType = folderType.includes(type) ? FileTypes.Folder : FileTypes.File;

                // Because of the same id in different levels, so we should set another uniq id for each tree node
                const id = `${catalogue.id}-${folderType.includes(type) ? 'folder' : 'file'}`;
                return new TreeNodeModel({
                    id,
                    name,
                    location: name,
                    fileType,
                    isLeaf: fileType === FileTypes.File,
                    data: catalogue,
                    icon: fileIcon(catalogue.taskType, source),
                    children: functionManagerService.get(id)?.children || [],
                });
            }

            case CATALOGUE_TYPE.TASK: {
                const fileType = folderType.includes(catalogue.type) ? FileTypes.Folder : FileTypes.File;

                // prevent same id between folder and file
                const id = fileType === FileTypes.File ? catalogue.id : `${catalogue.id}-folder`;
                return new TreeNodeModel({
                    id,
                    name: catalogue.name,
                    location: catalogue.name,
                    fileType,
                    icon: fileIcon(catalogue.taskType, source),
                    isLeaf: fileType === FileTypes.File,
                    data: catalogue,
                    children: molecule.folderTree.get(id)?.children || [],
                });
            }

            default:
                return undefined;
        }
    };

    private getServiceBySource = (source: CATALOGUE_TYPE) => {
        switch (source) {
            case CATALOGUE_TYPE.TASK:
                return molecule.folderTree;
            case CATALOGUE_TYPE.FUNCTION:
                return functionManagerService;
            case CATALOGUE_TYPE.RESOURCE:
                return resourceManagerService;
            default:
                return null;
        }
    };

    /**
     * @param node 更新当前 Node 节点的子节点，不改变 Node 节点
     */
    public loadTreeNode = async (
        node: Partial<Pick<CatalogueDataProps, 'id' | 'catalogueType'>>,
        source: CATALOGUE_TYPE
    ) => {
        const data = await this.getCatalogueViaNode(node);
        if (data) {
            const childrenNodes = <molecule.model.TreeNodeModel[]>(
                (data?.children?.map((child) => this.transformCatalogueToTree(child, source)) || []).filter(Boolean)
            );
            const service = this.getServiceBySource(source);
            const treeNode = service?.get(`${node.id!}-folder`);
            if (treeNode) {
                service?.update({
                    ...treeNode,
                    children: childrenNodes,
                });

                this.emit(CatalogueEventKind.onUpdate);
                return service?.get(treeNode.id);
            }
        }
    };

    public getRootFolder = (source: CATALOGUE_TYPE) => {
        const service = this.getServiceBySource(source);
        switch (source) {
            case CATALOGUE_TYPE.TASK:
            case CATALOGUE_TYPE.FUNCTION:
                return service?.getState().folderTree?.data?.[0];
            case CATALOGUE_TYPE.RESOURCE:
                return service?.getState().folderTree?.data?.[0]?.children?.[0];
            default:
                return undefined;
        }
    };

    public loadRootFolder = () => {
        this.getCatalogueViaNode({ id: 0 }).then((res) => {
            if (!res || !res.children) {
                return;
            }
            const { children } = res;

            const taskData = this.getRootFolderViaSource(children, CATALOGUE_TYPE.TASK);
            const resourceData = this.getRootFolderViaSource(children, CATALOGUE_TYPE.RESOURCE);
            const funcData = this.getRootFolderViaSource(children, CATALOGUE_TYPE.FUNCTION);

            // 更新资源目录树
            if (resourceData) {
                const resourceRoot = resourceData;
                const resourceNode = this.transformCatalogueToTree(resourceRoot, CATALOGUE_TYPE.RESOURCE)!;
                const childrenNodes =
                    resourceRoot.children?.map(
                        (child) => this.transformCatalogueToTree(child, CATALOGUE_TYPE.RESOURCE)!
                    ) || [];

                // set a root folder
                resourceNode.fileType = FileTypes.RootFolder;
                // put children to root folder
                resourceNode.children = childrenNodes;
                resourceManagerService.add(resourceNode);
            }

            // 更新函数目录树
            if (funcData) {
                const funcRoot = funcData;
                const functionNode = this.transformCatalogueToTree(funcRoot, CATALOGUE_TYPE.FUNCTION)!;
                const childrenNodes =
                    funcRoot.children
                        // there is a system function in the children node of root folder, we'd better to filter it
                        ?.filter((child) => child.name !== '系统函数')
                        .map((child) => this.transformCatalogueToTree(child, CATALOGUE_TYPE.FUNCTION)!) || [];

                // set a root folder
                functionNode.fileType = FileTypes.RootFolder;
                // put children to root folder
                functionNode.children = childrenNodes;
                functionManagerService.add(functionNode);

                // // sql 节点必存在 catalogueType，对所有的节点都求一遍子树
                // const SqlNodes =
                // 	funcRoot?.children?.filter((child: any) => child.catalogueType) || [];
                // SqlNodes.forEach((sqlNode) => {
                // 	loadTreeNode(sqlNode, CATELOGUE_TYPE.FUNCTION);
                // });
            }

            // 任务开发根目录
            if (taskData) {
                const taskRootFolder = taskData?.children?.[0];
                if (taskRootFolder) {
                    const taskNode = this.transformCatalogueToTree(taskRootFolder, CATALOGUE_TYPE.TASK)!;

                    taskNode.fileType = FileTypes.RootFolder;
                    molecule.folderTree.add(taskNode);

                    // 获取当前根目录的下级目录，确保打开 Explorer 有数据展示
                    this.loadTreeNode(taskRootFolder, CATALOGUE_TYPE.TASK);
                }
            }
        });
    };

    public onUpdate = (callback: () => void) => {
        this.subscribe(CatalogueEventKind.onUpdate, callback);
    };
}
