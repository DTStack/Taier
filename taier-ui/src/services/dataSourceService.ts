/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { IDataSourceProps } from '@/interface';
import { Component } from '@dtinsight/molecule/esm/react';
import Base64 from 'base-64';
import api from '@/api';

export interface IDataSourceState {
    dataSource: IDataSourceProps[];
}

interface IDataSourceService {
    getDataSource: () => IDataSourceProps[];
    reloadDataSource: () => void;
}

export default class DataSourceService extends Component<IDataSourceState> implements IDataSourceService {
    protected state: IDataSourceState = {
        dataSource: [],
    };

    constructor() {
        super();
        this.queryDataSource();
    }

    private queryDataSource = () => {
        api.getAllDataSource({}).then((res) => {
            if (res.code === 1) {
                const nextData: IDataSourceProps[] = ((res.data as IDataSourceProps[]) || []).map((ele) => {
                    const canConvertLinkJson =
                        ele.linkJson && !ele.linkJson.includes('{') && !ele.linkJson.includes('}');

                    return {
                        ...ele,
                        linkJson: canConvertLinkJson ? Base64.decode(ele.linkJson!) : ele.linkJson,
                    };
                });

                this.setState({
                    dataSource: nextData,
                });
            }
        });
    };

    getDataSource = () => {
        return this.state.dataSource || [];
    };

    reloadDataSource = () => {
        this.queryDataSource();
    };
}
