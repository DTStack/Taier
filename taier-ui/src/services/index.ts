import { container } from 'tsyringe';
import EditorActionBarService from './editorActionBarService';
import ExecuteService from './executeService';
import CatalogueService from './catalogueService';
import BreadcrumbService from './breadcrumbService';
import RightBarService from './rightBarService';
import TaskRenderService from './taskRenderService';
import DataSourceService from './dataSourceService';
import TaskParamsService from './taskParamsService';

const editorActionBarService = container.resolve(EditorActionBarService);
const executeService = container.resolve(ExecuteService);
const catalogueService = container.resolve(CatalogueService);
const breadcrumbService = container.resolve(BreadcrumbService);
const rightBarService = container.resolve(RightBarService);
const taskRenderService = container.resolve(TaskRenderService);
const dataSourceService = container.resolve(DataSourceService);
const taskParamsService = container.resolve(TaskParamsService);

export {
    editorActionBarService,
    catalogueService,
    executeService,
    breadcrumbService,
    rightBarService,
    taskRenderService,
    dataSourceService,
    taskParamsService,
};
