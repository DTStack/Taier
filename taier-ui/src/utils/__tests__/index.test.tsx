import { TASK_STATUS } from '@/constant';
import { taskRenderService } from '@/services';
import '@testing-library/jest-dom';
import moment from 'moment';
import { history } from 'umi';
import {
    checkExist,
    convertObjToNamePath,
    convertParams,
    copyText,
    createElement,
    createSeries,
    createSQLProposals,
    deleteCookie,
    disableRangeCreater,
    filterComments,
    filterSql,
    formatDateTime,
    getColumnsByColumnsText,
    getCookie,
    getPlus,
    getTenantId,
    getTodayTime,
    getUserId,
    getVertexStyle,
    goToTaskDev,
    isValidFormatType,
    pickByTruly,
    prettierJSONstring,
    randomId,
    removePopUpMenu,
    renderCharacterByCode,
    replaceStrFormIndexArr,
    splitByKey,
    splitSql,
    toArray,
    utf16to8,
    utf8to16,
    visit,
} from '..';

jest.mock('@/components/customDrawer', () => {
    return {
        updateDrawer: jest.fn(),
    };
});

jest.mock('umi', () => {
    return {
        history: { push: jest.fn() },
    };
});

jest.mock('@/services', () => {
    return {
        taskRenderService: {
            openTask: jest.fn(),
        },
    };
});

describe('utils/index', () => {
    beforeEach(() => {
        jest.clearAllMocks();
        document.body.innerHTML = '';
    });

    it('Should Get the time', () => {
        const today = getTodayTime();
        expect(today).toHaveLength(2);
        expect(today[0]).toBeInstanceOf(moment);
        expect(today[1]).toBeInstanceOf(moment);
        expect(today[0].format('HH:mm:ss')).toBe('00:00:00');
        expect(today[1].format('HH:mm:ss')).toBe('23:59:59');
    });

    it('Should Get the Cookie', () => {
        document.cookie = 'test=abc;';
        document.cookie = 'test2=zxc;';
        expect(getCookie('test')).toBe('abc');
        expect(getCookie('test2')).toBe('zxc');
        expect(getCookie('test3')).toBe(null);
    });

    it('Should Delete the Cookie', () => {
        expect(getCookie('test')).toBe('abc');

        deleteCookie('test');
        expect(getCookie('test')).toBe(null);
    });

    it('Should Format the DateTime', () => {
        expect(formatDateTime('2022-06-01')).toBe('2022-06-01 00:00:00');
        expect(formatDateTime(new Date('2022-06-01').valueOf())).toBe('2022-06-01 00:00:00');
        expect(formatDateTime(new Date('2022-06-01'))).toBe('2022-06-01 00:00:00');
    });

    it('Should Get the Existension', () => {
        expect(checkExist(undefined)).toBeFalsy();
        expect(checkExist(null)).toBeFalsy();
        expect(checkExist('')).toBeFalsy();
        expect(checkExist({})).toBeTruthy();
    });

    it('Should Filter Comments In SQL', () => {
        expect(filterComments('show tables;-- name test')).toBe('show tables; ');
        expect(filterComments('-- select test from A where id = "taier";')).toBe(' ');
        expect(filterComments('-- name test\nshow tables;-- name test')).toBe(' \nshow tables; ');
        expect(filterComments('/* select test from A where id = "taier"; */')).toBe(' ');
        expect(filterComments('select test from A where id = "taier";-- name test')).toBe(
            'select test from A where id = "taier"; '
        );

        // Error
        expect(filterComments('show tables";-- name test')).toBe('show tables";-- name test');
        // Error
        expect(filterComments('show tables/*;-- name test')).toBe('show tables/*;-- name test');
    });

    it('Should Replace String By Index Array', () => {
        expect(replaceStrFormIndexArr('test', 'abc', [{ begin: 0, end: 2 }])).toBe('abct');
        expect(replaceStrFormIndexArr('test', 'abc', [])).toBe('test');
    });

    it('Should Split Sqls', () => {
        expect(splitSql('')).toEqual('');
        expect(splitSql('show tables')).toEqual(['show tables']);
        expect(splitSql('show tables;')).toEqual(['show tables']);
        expect(splitSql('show tables;select * from tableA where id = "1";')).toEqual([
            'show tables',
            'select * from tableA where id = "1"',
        ]);
    });

    it('Should Filter Comments And Remove Spaces', () => {
        const sql = 'show tables;-- name test';
        expect(filterSql(sql)).toEqual(['show tables']);
    });

    it('Should Get Tenant Id From Cookie', () => {
        document.cookie = 'tenantId=1;';
        expect(getTenantId()).toBe('1');
        deleteCookie('tenantId');
    });

    it('Should Get User Id From Cookie', () => {
        document.cookie = 'userId=1;';
        expect(getUserId()).toBe('1');
        deleteCookie('userId');
    });

    it('Should Remove Popup Menu from Document', () => {
        jest.useFakeTimers();
        const dom = document.createElement('div');
        dom.classList.add('mxPopupMenu');
        dom.style.visibility = 'visibility';
        document.body.appendChild(dom);
        expect(dom.style.visibility).toBe('visibility');

        removePopUpMenu();

        jest.advanceTimersByTime(500);

        expect(dom.style.visibility).toBe('hidden');
    });

    it('Should render different status', () => {
        expect(getVertexStyle(TASK_STATUS.AUTO_CANCEL)).toBe(
            'whiteSpace=wrap;fillColor=var(--badge-cancel-background);strokeColor=var(--badge-cancel-border);'
        );

        expect(getVertexStyle(TASK_STATUS.COMPUTING)).toBe(
            'whiteSpace=wrap;fillColor=var(--badge-pending-background);strokeColor=var(--badge-pending-border);'
        );

        expect(getVertexStyle(TASK_STATUS.FINISHED)).toBe(
            'whiteSpace=wrap;fillColor=var(--badge-finished-background);strokeColor=var(--badge-finished-border);'
        );

        expect(getVertexStyle(TASK_STATUS.RUNNING)).toBe(
            'whiteSpace=wrap;fillColor=var(--badge-running-background);strokeColor=var(--badge-running-border);'
        );

        expect(getVertexStyle(TASK_STATUS.PARENT_FAILD)).toBe(
            'whiteSpace=wrap;fillColor=var(--badge-failed-background);strokeColor=var(--badge-failed-border);'
        );

        // default
        expect(getVertexStyle(99999)).toBe(
            'whiteSpace=wrap;fillColor=var(--badge-common-background);strokeColor=var(--badge-common-border);'
        );
    });

    it('Should prettier JSON string', () => {
        expect(prettierJSONstring(JSON.stringify({ a: { b: 1 }, c: JSON.stringify({ a: 1 }) }))).toBe(
            '{\n  "a": {\n    "b": 1\n  },\n  "c": {\n    "a": 1\n  }\n}'
        );

        expect(prettierJSONstring(';;')).toBe(';;');
    });

    it('Should have a randomId with 13 length', () => {
        expect(typeof randomId()).toBe('number');
        expect(randomId().toString().length).toBe(13);
    });

    it('Should support copyText', () => {
        const mockFn = jest.fn();
        Object.defineProperty(navigator, 'clipboard', {
            configurable: true,
            value: { writeText: mockFn },
        });

        copyText('test');
        expect(mockFn).toBeCalled();
        expect(mockFn.mock.calls[0][0]).toBe('test');

        Reflect.deleteProperty(navigator, 'clipboard');

        // Expect call execCommand without navigator.clipboard.writeText
        mockFn.mockClear();
        Object.defineProperty(document, 'execCommand', {
            value: mockFn,
        });
        copyText('test');
        expect(mockFn).toBeCalled();
        Reflect.deleteProperty(document, 'execCommand');
    });

    it('Should have a series of number', () => {
        expect(createSeries(5)).toEqual([1, 2, 3, 4, 5]);
    });

    it('Should parse columns', () => {
        expect(getColumnsByColumnsText('id int')).toEqual([{ field: 'id', type: 'int' }]);
        expect(getColumnsByColumnsText('id int\nage int\n\n\n\n')).toEqual([
            { field: 'id', type: 'int' },
            { field: 'age', type: 'int' },
        ]);
    });

    it('Should render Character', () => {
        expect(renderCharacterByCode(8)).toBe('âŒ«');
    });

    it('Should get an Array', () => {
        expect(toArray('test')).toEqual(['test']);
        expect(toArray(['test'])).toEqual(['test']);
    });

    it('Should convert params into values', () => {
        expect(
            convertParams({ sourceId: '{{ form#a.b }}', targetId: '{{ form#a.b#toArray }}' }, { a: { b: 1 } })
        ).toEqual({
            sourceId: 1,
            targetId: [1],
        });
    });

    it('Should get an calculated value', () => {
        expect(getPlus({ a: { b: [{ c: 1, value: 100 }] } }, '{{a.b#find.c}}', 100)).toBe(1);

        expect(getPlus({ a: { b: 2 } }, '{{a.b}}')).toBe(2);
    });

    it("Should convert to form's name path", () => {
        expect(convertObjToNamePath({ a: { b: 1 } })).toEqual([['a', 'b'], 1]);
    });

    it('Should support to visit tree nodes by condition', () => {
        const tree = {
            type: 1,
            value: 1,
            children: [
                {
                    type: 'number',
                    value: 1,
                },
                {
                    type: 'number',
                    value: 1,
                },
                {
                    type: 'number',
                    value: 1,
                },
                {
                    type: 'object',
                    value: 2,
                    children: [
                        {
                            type: 'object',
                            value: 1,
                            children: [
                                {
                                    type: 'number',
                                    value: 2,
                                },
                            ],
                        },
                    ],
                },
            ],
        };
        visit(
            tree,
            (item) => item.value === 1,
            (item) => {
                item.visited = true;
            }
        );

        expect(tree).toEqual({
            type: 1,
            value: 1,
            children: [
                {
                    type: 'number',
                    value: 1,
                    visited: true,
                },
                {
                    type: 'number',
                    value: 1,
                    visited: true,
                },
                {
                    type: 'number',
                    value: 1,
                    visited: true,
                },
                {
                    type: 'object',
                    value: 2,
                    children: [
                        {
                            type: 'object',
                            value: 1,
                            visited: true,
                            children: [
                                {
                                    type: 'number',
                                    value: 2,
                                },
                            ],
                        },
                    ],
                },
            ],
        });
    });

    it('Should filter the falsy value in object', () => {
        expect(pickByTruly({ a: 0, b: '', c: undefined, d: null })).toEqual({ a: 0, b: '' });
    });

    it('Should pick up object by keys', () => {
        expect(splitByKey({ a: 1, b: 2, c: 3 }, ['a', 'b'])).toEqual({
            obj1: {
                a: 1,
                b: 2,
            },
            obj2: {
                c: 3,
            },
        });
    });

    it('Should distinguish the validation of formate type', () => {
        expect(isValidFormatType('')).toBe(false);
        expect(isValidFormatType('string')).toBe(true);
        expect(isValidFormatType('STRING')).toBe(true);
        expect(isValidFormatType('varChar')).toBe(true);
        expect(isValidFormatType('varChar2')).toBe(true);
        expect(isValidFormatType('abc')).toBe(false);
    });

    it('Should have completion for sql', () => {
        window.fetch = jest
            .fn()
            .mockResolvedValueOnce({
                json: () => Promise.resolve(require('../../../public/assets/keywords.json')),
            })
            .mockResolvedValueOnce({
                json: () => Promise.resolve(require('../../../public/assets/functions.json')),
            });
        const completions = createSQLProposals({
            startLineNumber: 0,
            startColumn: 0,
            endLineNumber: 0,
            endColumn: 0,
        });

        expect(completions).resolves.toMatchSnapshot();
    });

    it('disableRangeCreater', () => {
        expect(disableRangeCreater(null, null, 'hour')).toEqual([]);

        // [0,1,2,3,4,...,23]
        const hoursRange = new Array(24).fill(1).map((_, idx) => idx);
        // [0,1,2,3,4,...,59]
        const minutesRange = new Array(60).fill(1).map((_, idx) => idx);
        // [0,1,2,3,4,...,59]
        const secondsRange = new Array(60).fill(1).map((_, idx) => idx);

        // Range span a day
        expect(
            disableRangeCreater(moment('2013-02-08', 'YYYY-DD-MM'), moment('2013-02-09', 'YYYY-DD-MM'), 'hour')
        ).toEqual(hoursRange.slice(1));

        expect(
            disableRangeCreater(
                moment('2013-02-08 09:30', 'YYYY-DD-MM HH:mm'),
                moment('2013-02-08 10:30', 'YYYY-DD-MM HH:mm'),
                'hour'
            )
        ).toEqual(hoursRange.slice(11));

        expect(
            disableRangeCreater(
                moment('2013-02-08 09:30', 'YYYY-DD-MM HH:mm'),
                moment('2013-02-08 10:30', 'YYYY-DD-MM HH:mm'),
                'hour',
                true
            )
        ).toEqual(hoursRange.slice(0, 9));

        // Returns [] since different hour
        expect(
            disableRangeCreater(
                moment('2013-02-08 09:30', 'YYYY-DD-MM HH:mm'),
                moment('2013-02-08 10:35', 'YYYY-DD-MM HH:mm'),
                'minute'
            )
        ).toEqual([]);

        expect(
            disableRangeCreater(
                moment('2013-02-08 09:30', 'YYYY-DD-MM HH:mm'),
                moment('2013-02-08 9:35', 'YYYY-DD-MM HH:mm'),
                'minute'
            )
        ).toEqual(minutesRange.slice(36));

        // Returns [] since different hour
        expect(
            disableRangeCreater(
                moment('2013-02-08 09:30:26', 'YYYY-DD-MM HH:mm:ss'),
                moment('2013-02-08 10:30:26', 'YYYY-DD-MM HH:mm:ss'),
                'second'
            )
        ).toEqual([]);

        // Returns [] since different minute
        expect(
            disableRangeCreater(
                moment('2013-02-08 09:30:26', 'YYYY-DD-MM HH:mm:ss'),
                moment('2013-02-08 09:31:26', 'YYYY-DD-MM HH:mm:ss'),
                'second'
            )
        ).toEqual([]);

        expect(
            disableRangeCreater(
                moment('2013-02-08 09:30:26', 'YYYY-DD-MM HH:mm:ss'),
                moment('2013-02-08 09:30:50', 'YYYY-DD-MM HH:mm:ss'),
                'second'
            )
        ).toEqual(secondsRange.slice(51));
    });

    it('Should convert utf-8 to utf-16', () => {
        expect(utf8to16('test')).toBe('test');
        expect(utf8to16('@Ã€Ã¤Ã¥Ã¦Ã§Ã¨Ã©')).toBe('@$å¦§è©€');
        expect(utf8to16(1 as any)).toBe(1);
    });

    it('Should convert utf-16 to utf-8', () => {
        expect(utf16to8(1 as any)).toBe(1);
        expect(utf16to8('Ã€Ã¤Ã¥Ã¦Ã§Ã¨Ã©')).toBe('Ã€Ã¤Ã¥Ã¦Ã§Ã¨Ã©');
        expect(utf16to8('aÂ¡à¤„')).toBe('aÃ‚Â¡Ã Â¤Â„');
    });

    it('Should support to goToTaskDev', () => {
        goToTaskDev({ id: 1 });

        expect(taskRenderService.openTask).toBeCalledWith({ id: '1' });
        expect(history.push).toBeCalledWith({ query: {} });
    });

    it('Should create a dom avoid duplicated element', () => {
        expect(document.body.childElementCount).toBe(0);

        createElement({});
        expect(document.body.childElementCount).toBe(1);

        createElement({ className: 'test' });
        expect(document.querySelector('.test')).toBeInTheDocument();
    });
});
