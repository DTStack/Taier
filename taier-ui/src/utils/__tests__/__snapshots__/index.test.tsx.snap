// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`utils/index Should have completion for sql 1`] = `
Array [
  Object {
    "documentation": "The SELECT statement is used to select data from a database.",
    "insertText": "SELECT",
    "kind": 0,
    "label": "SELECT",
    "range": Object {
      "endColumn": 0,
      "endLineNumber": 0,
      "startColumn": 0,
      "startLineNumber": 0,
    },
  },
  Object {
    "documentation": "The CREATE TABLE statement is used to create a new table in a database.",
    "insertText": "CREATE TABLE",
    "kind": 0,
    "label": "CREATE TABLE",
    "range": Object {
      "endColumn": 0,
      "endLineNumber": 0,
      "startColumn": 0,
      "startLineNumber": 0,
    },
  },
  Object {
    "documentation": "The WHERE clause is used to filter records.",
    "insertText": "WHERE",
    "kind": 0,
    "label": "WHERE",
    "range": Object {
      "endColumn": 0,
      "endLineNumber": 0,
      "startColumn": 0,
      "startLineNumber": 0,
    },
  },
  Object {
    "documentation": "The AND operator is used to filter records based on more than one condition, which displays a record if all the conditions separated by AND are TRUE.",
    "insertText": "AND",
    "kind": undefined,
    "label": "AND",
    "range": Object {
      "endColumn": 0,
      "endLineNumber": 0,
      "startColumn": 0,
      "startLineNumber": 0,
    },
  },
  Object {
    "documentation": "The OR operator is used to filter records based on more than one condition, which displays a record if any of the conditions separated by OR is TRUE.",
    "insertText": "OR",
    "kind": undefined,
    "label": "OR",
    "range": Object {
      "endColumn": 0,
      "endLineNumber": 0,
      "startColumn": 0,
      "startLineNumber": 0,
    },
  },
  Object {
    "documentation": "The NOT operator displays a record if the condition(s) is NOT TRUE.",
    "insertText": "NOT",
    "kind": undefined,
    "label": "NOT",
    "range": Object {
      "endColumn": 0,
      "endLineNumber": 0,
      "startColumn": 0,
      "startLineNumber": 0,
    },
  },
  Object {
    "documentation": "The ORDER BY keyword is used to sort the result-set in ascending or descending order. The ORDER BY keyword sorts the records in ascending order by default. To sort the records in descending order, use the DESC keyword.",
    "insertText": "ORDER BY",
    "kind": 0,
    "label": "ORDER BY",
    "range": Object {
      "endColumn": 0,
      "endLineNumber": 0,
      "startColumn": 0,
      "startLineNumber": 0,
    },
  },
  Object {
    "documentation": "The INSERT INTO statement is used to insert new records in a table.",
    "insertText": "INSERT INTO",
    "kind": 0,
    "label": "INSERT INTO",
    "range": Object {
      "endColumn": 0,
      "endLineNumber": 0,
      "startColumn": 0,
      "startLineNumber": 0,
    },
  },
  Object {
    "documentation": "A field with a NULL value is a field with no value.",
    "insertText": "NULL",
    "kind": undefined,
    "label": "NULL",
    "range": Object {
      "endColumn": 0,
      "endLineNumber": 0,
      "startColumn": 0,
      "startLineNumber": 0,
    },
  },
  Object {
    "documentation": "The UPDATE statement is used to modify the existing records in a table.",
    "insertText": "UPDATE",
    "kind": 0,
    "label": "UPDATE",
    "range": Object {
      "endColumn": 0,
      "endLineNumber": 0,
      "startColumn": 0,
      "startLineNumber": 0,
    },
  },
  Object {
    "documentation": "The DELETE statement is used to modify the existing records in a table.",
    "insertText": "DELETE",
    "kind": 0,
    "label": "DELETE",
    "range": Object {
      "endColumn": 0,
      "endLineNumber": 0,
      "startColumn": 0,
      "startLineNumber": 0,
    },
  },
  Object {
    "documentation": "The INSERT INTO statement is used to insert new records in a table.",
    "insertText": "INSERT INTO \${1:tableName} VALUES(\${2:value});",
    "insertTextRules": 0,
    "kind": undefined,
    "label": "INSERT:snippet",
    "range": Object {
      "endColumn": 0,
      "endLineNumber": 0,
      "startColumn": 0,
      "startLineNumber": 0,
    },
  },
  Object {
    "documentation": "The CREATE TABLE statement is used to create a new table in a database.",
    "insertText": "CREATE TABLE \${1:table_name} (
	\${2:column} \${3:datatype}
);",
    "insertTextRules": 0,
    "kind": undefined,
    "label": "CREATE:snippet",
    "range": Object {
      "endColumn": 0,
      "endLineNumber": 0,
      "startColumn": 0,
      "startLineNumber": 0,
    },
  },
  Object {
    "documentation": "The SELECT statement is used to select data from a database.",
    "insertText": "SELECT \${1:column} FROM \${2:table_name};",
    "insertTextRules": 0,
    "kind": undefined,
    "label": "SELECT:snippet",
    "range": Object {
      "endColumn": 0,
      "endLineNumber": 0,
      "startColumn": 0,
      "startLineNumber": 0,
    },
  },
  Object {
    "documentation": "show tables;",
    "insertText": "show tables;",
    "insertTextRules": 0,
    "kind": undefined,
    "label": "SHOW:snippet",
    "range": Object {
      "endColumn": 0,
      "endLineNumber": 0,
      "startColumn": 0,
      "startLineNumber": 0,
    },
  },
]
`;
